docs/smart-contracts/capabilities.md

```diff
@@ -0,0 +1,279 @@
+- One or more _parameters_ to specify properties or conditions for the capability to be granted.
+- A _predicate function_ to determine whether the capability is acquired or not.
+
+For example, the following sample code defines an ALLOW_ENTRY capability:
+
+```pact
+(defcap ALLOW_ENTRY (user-id:string)
+  "Govern entry operation."
+  (with-read table user-id
+    { "guard" := guard, "active" := active }
+    (enforce-guard guard)
+    (enforce active "Only active users allowed entry"))
+)
+```
+
+In this example:
+
+- `ALLOW_ENTRY` is the _name_ or _domain_ of the capability. 
+
+- `user-id` is a _parameter_. 
+  Together, the name and parameter attributes form the _specification_ of the capability. 
+  Therefore, `(ALLOW_ENTRY 'dave)` and `(ALLOW_ENTRY 'carol)` describe separate capabilities.
+
+- The remaining body implements the _predicate function_ for the capability. 
+  In its simplest form, the predicate function accesses whatever data it needs to perform the tests necessary to protect against improper granting of the capability. 
+  The body can also do more, including import or _compose_ additional capabilities or modify database state. This might be used to ensure a capability cannot be granted ever again after the first time it is acquired, for example.
+
+Because `defcap` declarations both _specify_ a domain of capability instances, and _implement_ the guard function, the capability grant doesn't need to be called when invoked in `with-capability` or `require-capability` function calls.
+You can only grant a capability within the module where it's defined.
+
+### Acquire a capability
+
+You can grant a capability to a function by using the `(with-capability)` function.
+For example, to acquire the `ALLOW_ENTRY` capability, you would call the `with-capability` function like this:
+
+```pact
+(defun enter (user-name)
+  (with-capability (ALLOW_ENTRY user-name)
+    (do-entry user-name)            ;; call "protected" function
+    (update-entry-status user-name) ;; update database
+  )
+  (record-audit "ENTRY" user-name)  ;; some "unsafe" operation
+)
+```
+
+When capabilities are granted, they are installed into the Pact environment for the scope of the call to the `with-capability` function.
+After the code exits the call to the `with-capability` function, the capability is removed. 
+This scoping prevents duplicate testing of the predicate.
+Capabilities that have already been acquired and are in-scope are not re-evaluated.
+
+### Require a capability
+
+You can protect sensitive code with the `(require-capability)` function. 
+With the  `(require-capability)` function, code can demand that a capability be already granted, that is, make no attempt to acquire the ticket, but fail if it was not acquired somewhere else.
+To demand or _require_ the capability, you would call the `require-capability` function like this:
+
+```pact
+(defun do-entry (user-name)
+  (require-capability (ALLOW_ENTRY user-name))
+  ...
+)
```

Comment:
it's important to note in this that `require-capability` doesn't scope a body, so the position at which it's inserted actually does affect the semantics of a function call and what operations happen first, before requirement. I don't think this necessarily means that the transaction will fail depending on the position, but you might get weird error messages in different places, and it's best practice to just add it at the top of the function